### ISSUE TITLE
Implement the W3C TDM Reservation Protocol and enable a more standard opt-out mechanism

### ISSUE THREAD

**Issue by llemeurfr** [rom1504/img2dataset#308]
The solution which was chosen by the author after a heated discussion in #293 was to support an opt-out expressed in http headers, via well know values "noindex" and "noimageindex" plus the ad-hoc values "noai" and "noimageai". 

This is already a good move: in Europe, any crawler associated with TDM and AI technologies MUST support opt-out, as stipulated by the European DSM Directive. You'll get more information [here](https://w3c.github.io/tdm-reservation-protocol/) about that legal requirement. Because this soft is gathering images available for AI training, it should not integrate in its dataset images for which authors have decided an opt-out. 

But "noai" and "noimageia" are not well known tokens (even if you're not alone trying them), there is nothing standard in them so far. And robots.txt is not only about http headers. Directives can be in a file stored at the root of the web site (and as html meta, but this is not interesting here). Therefore your move does not really help the community establishing trusted relationships between AI solutions and content providers (which is a requirement if you want content providers to see AI actors as partners, not enemies). 
 
For this reason, a  W3C Community Group constituted of content providers and TDM actors decided to create an open specification two years ago, and released [this specification called TDMRep](https://www.w3.org/2022/tdmrep/) (for TDM Reservation Protocol). The [home page of the group is there](https://www.w3.org/community/tdmrep/); 42 participants. 

For those wondering, [this specification also covers AI solutions](https://w3c.github.io/tdm-reservation-protocol/docs/tdm-meaning.html). And this group [didn't use robots.txt](https://w3c.github.io/tdm-reservation-protocol/docs/robots.html) for clear reasons. 

Adding the support of a new property in the http header, called "tdm-reservation", and filtering images if its value is 1 (number)  is a no-brainer. Adding the support of a JSON file named tdmrep.json, hosted in the /.well-known repository of the Web server on which the image is stored, is a bit more complex, but still easy in Python (it is identical to the processing of the robots.txt file); and its is mandatory even if less performant.

**Comment by llemeurfr** [comment]
Hi @maathieu, 
To what extent is `X-Robots-Tag: noai` more standard than `tdm-reservation: 1`?  A Web server must in both cases be tuned to generate the property. 

The evolution of [is_disallowed()]( https://github.com/rom1504/img2dataset/blob/21c297b089cf7bc30480825af83bf1859862f70d/img2dataset/downloader.py#LL20C1-L34C17) is far from complex. Would it help if contributors propose a PR? 

The main complexity will in any case be to handle a robots.txt or tdmrep.json file.

Note: the W3C Community Group is waiting for more feedback from implementers before going through the W3C Recommendation route.

**Comment by rom1504** [comment]
Adding a new header in the set of disallowed ones seems indeed fine.
Note img2dataset did not introduce the noai and noimageai tags, there were suggested by DeviantArt, see #218 
It would be great to see this standardized instead of each website creating their own.

As for tdmrep.json, i am wondering if you considered specifying in the response headers that this file exists? If that were the case, it would be possible to check this file only for the (initially small minority) of websites providing it.

**Comment by llemeurfr** [comment]
> It would be great to see a set of opt out headers standardized instead of each website creating their own.

YES, this is why we would all like to have a worldwide standard for TDM and AI opt-out (and because a raw opt-out is not great for the future of AI, we're trying to allow deals to be made). 

> As for tdmrep.json, i am wondering if you considered specifying in the response headers that this file exists?

No, we didn't consider that. A simple HEAD on tdmrep.json that responds with a 404 is not a huge lack of performance. PS: You will have to parse robots.txt if you want to check its rules, which takes much more time.

**Comment by DavidJCobb** [comment]
> To what extent is `X-Robots-Tag: noai` more standard than `tdm-reservation: 1`?

*Neither* of them is "more standard." TDMRep is a draft specification, it (by necessity) explicitly states that it's not a W3C standard, and it only has one author and editor: you. I counted participants in the mailing by hand and saw 20 unique names, which isn't even enough for the "There are dozens of us! Dozens!" meme.

Using your not-a-standard wouldn't make the owner of this repo any less inconsiderate or any less of a disingenuous clown.

**Comment by maathieu** [comment]
@DavidJCobb , good point, that was a nice strawman argument from @llemeurfr . Robots.txt does not require any server tuning. Just place the file in the root directory. The scraper will download the file once, then compare all the URLs it wishes to scrape to the data placed in robots.txt . If there is no match, then it can download. If there is a match, then no download. There is no need to invent something new and convoluted to replace this established practice. A scraper is not different from a search engine spider. Google has also been doing "AI" for decades, and it has respected robots.txt on webmaster's websites. Please be good netizens. It's not because AI is the shiny new thing that all established practices must be abandoned.
